[CS2103T-W12-2] THRIFT

limerencee opened the pull request:
@kensonoenqh @krusagiz @lightz96 @lye-jw

limerencee committed:
Remove the codacy badge from README

lightz96 committed:
Removes SE-EDU branding (#46)

* Removes SE-EDU branding

* Updates UserGuide and DeveloperGuide

* Updates travis ci badge

* Updates appveyor badge

* Updates appveyor badge

* Updates coveralls badge

* Adds netlify badge and removes learning outcomes link

limerencee committed:
Changed more documents context to THRIFT (#51)

* Changed documents context to THRIFT

* Modified UserGuide and AboutUs

Improved wording clarity in UserGuide and refined my role in AboutUs.

* Updated UserGuide

Modified wording and formatting for 'Delete' command.

* Updated UserGuide

lightz96 committed:
Update documentation (#59)

* Adds MIT license and Updates badges logo

* Updates AboutUs and creates dummy portfolio

* Update Ui image

* Update user guide

* Update aboutus and user guide

* Update developer guide and minor changes to user guide and aboutus

* Change to LF

* Add newline at EOF for my portfolio

lye-jw committed:
Update AboutUs.adoc and UserGuide.adoc (Jian Wen) (#56)

* Update AboutUs.doc (Jian Wen)

* Update UserGuide.adoc

* Update lye-jw.adoc Portfolio file

* Refine UserGuide

* Fix checkStyle issue

kensonoenqh committed:
Updated developer guide with one use case (#64)

* Updated developer guide with one use case

* Commit the image file that i forgot.

kensonoenqh committed:
Updated UserGuide.adoc and AboutUs.adoc (#63)

* Updated UserGuide.adoc and AboutUs.adoc

* removed remaining budget line in userguide as per jiahaos suggestion.

* Fixed typo, also edited help command because i edited it by mistake before.

* Fixed typo from resolving conflict in UserGuide

* Added one new line to end of userguide

* Improved user guide for format

limerencee committed:
Updated DeveloperGuide to include use cases for features that I am working on

limerencee committed:
Updated DG appendix

krusagiz committed:
Modified .gitattributes (#70)

limerencee committed:
Mass Codebase Refactor (#62)

* Changed documents context to THRIFT

* Modified UserGuide and AboutUs

Improved wording clarity in UserGuide and refined my role in AboutUs.

* Renamed 'model' package to 'transaction' and began refactoring

Added Transaction related classes and a new utility class which contains functions for converting currencies around.

* Renamed package from 'seedu.address' to 'thrift'

* Added tests for Transaction class attributes

* Mass refactor to change Person class to Transaction class

Also added Income and Expense classes, and replaced all instances of Person class references to either Transaction, Expense or Income class. This commit passes the JUnit test cases.

* Initial refactor complete

Application is able to run and able to add expense

limerencee committed:
Merge branch 'master' of https://github.com/AY1920S1-CS2103T-W12-2/main into update-dg

krusagiz committed:
Merge pull request #72 from limerencee/update-dg

Updated DeveloperGuide to include use cases for features that I am working on

krusagiz committed:
Added photo, Modified AboutUs and UserGuide (#69)

* Added photo, Modified AboutUs and UserGuide

* Fixed line endings

* Added photo, Modified AboutUs and UserGuide

* Fixed line endings

* Added some whitespace to AboutUs

* Changed format for convert and re-added tags lost to conflict

* Changed prefix for convert command

* Changed variables names and prefixes of convert command

limerencee committed:
Changed command syntax in UG for 'add expense' and 'add income' into one word (#74)

* Changed documents context to THRIFT

* Modified UserGuide and AboutUs

Improved wording clarity in UserGuide and refined my role in AboutUs.

* Renamed 'model' package to 'transaction' and began refactoring

Added Transaction related classes and a new utility class which contains functions for converting currencies around.

* Renamed package from 'seedu.address' to 'thrift'

* Added tests for Transaction class attributes

* Mass refactor to change Person class to Transaction class

Also added Income and Expense classes, and replaced all instances of Person class references to either Transaction, Expense or Income class. This commit passes the JUnit test cases.

* Initial refactor complete

Application is able to run and able to add expense

* Updated UG syntax for adding transactions

'add expense' -> 'add_expense' and 'add income' -> 'add_income'

* Fixed typo in UG

lightz96 committed:
Refactor code and comments from Person to Transaction, Address Book to Thrift (#76)

* Follow up on the comments mentioned in PR #62

* Remove Person class

* Removes Address class

* Removes Phone class

* Removes Name class

* Removes Email class

* Validate the description field in JsonAdaptedTransaction

* Rename to ThriftBuilder

* Rename to JsonSerializableThrift

* Rename from JsonAddressBookStorage to JsonThriftStorage

* Rename AddressBookStorage to ThriftStorage

* Rename AddressBookParser to ThriftParser

* Rename AddressBook to Thrift

* Change ordering of imports

* Change ordering of imports

* Change ordering of imports for tests

* Rename ReadOnlyAddressBook to ReadOnlyThrift

* Modify the JsonSerializableAddressBookTest folder name

* Rename the tests in test/data

* Rename to Thrift for the logic portion

* Update UserPrefs to use Thrift instead of Address book

* Update the userguide url in the app

* Update the naming for UI

* Rename withDescription() to withTransaction in ThriftBuilder

* Rename to Thrift

* Rename to Transaction instead of Person

* Update coding standards

* Rename Person to Transaction, Address Book to Thrift in all JUnit tests

* Rename to thrift in javadoc

* Update the icon filepath

* Added JUnit test for Messages.java

lye-jw committed:
Add on Update and Clone use cases in DeveloperGuide (#77)

lye-jw committed:
Add test cases for update and clone in DeveloperGuide (#78)

krusagiz committed:
Added use cases to DG and added logo (#79)

* Added Use cases to DG and added logo

- Logo files added to repo
- Logo added to DG and UG

* Edited DeveloperGuide.adoc and logos

- SVG versions of the logos are now included

* Added SVG to .gitattributes

* Added Illustrator file back to repo

* Added my test cases to the DG

* Removed an unnecessary period

limerencee committed:
Merge branch 'master' of https://github.com/nus-cs2103-AY1920S1/addressbook-level3 into fetch-from-address

kensonoenqh committed:
Kenson-DG (#80)

* Updated developer guide with one use case

* Commit the image file that i forgot.

* Added my test cases for listing transactions.

* Resolved the rest of the conflicts i missed out, and edited tagging a little (seems like some error conflict)

* Edited a few typos and errors.

* Remove monthly expenses

* added +s to follow formatting

* removed fullstop

limerencee committed:
Merge branch 'master' of https://github.com/AY1920S1-CS2103T-W12-2/main into fetch-from-address

krusagiz committed:
Merge pull request #81 from limerencee/fetch-from-address

Pulled new changes from AB3 repo

limerencee committed:
Edited UGDG to ensure consistency (#82)

* Edited UGDG to ensure consistency.

* Corrected casing for 'Transaction' to 'transaction'

lightz96 committed:
Update Developer Guide and User Guide (#83)

* Removes duplicate commands in command summary and updates the link for undoable commands

* Adds test cases in DG

* Adds one more test case for redo command

* Update formatting for redo and undo test cases

Aulud commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
-* can type fast
-* prefers typing over mouse input
-* is reasonably comfortable using CLI apps
+* money-conscious NUS students
```
As a suggestion, you can consider widening the scope of the product if the functionality can be extended to a wider target audience without much additional hassle, e.g. to students in general.

JohnNzj reviewed (commented):
Format: Unclear what "MSS = Extensions" means in Use Cases
Use cases format duplicated, can include use cases with more steps and variations with more specific details instead of using "invalid parameter"

Aulud commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
 Priorities: High (must have) - `* * \*`, Medium (nice to have) - `* \*`, Low (unlikely to have) - `*`
 
-[width="59%",cols="22%,<23%,<25%,<30%",options="header",]
+[width="100%",cols="5%,<25%,<30%,<40%",options="header",]
```
Some of the user stories seem similar and are excessively specific on a per-user basis, without specifying any extra features expected of the application. For example, user story numbers 10 and 17 both relate to viewing a common expense tracker report, and their benefit seems a bit contrived.

Aulud commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
+. A user with above average typing speed for <<regular-english-text, regular English text>> (i.e. not code, not system admin commands) should be able to accomplish most of the tasks faster using commands than using the mouse.
+. THRIFT should be able to run with or without internet connection.
+. THRIFT should work for a single user only.
+. THRIFT should not require user to install.
```
It is possible to clarify this as 'THRIFT is portable and does not require an installer.'

Aulud commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
-Pros:
+[[regular-english-text]] Regular English Text::
+Stands for text with ordinary english grammar structures and vocabulary generally used by the public.
+It excludes syntax related to programming and <<system-administration, system administration>>.
```
This exclusion is not necessary; the first sentence alone is sufficient. This also removes the need to further clarify the definition of 'System administration' further in the Glossary, and makes the Glossary more friendly to non-technical users.

Aulud commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
+|`*`
+|dota addict who always buy arcanas and battle pass levels
+|track my spending
+|afford to buy more battle pass levels
```
This is an example of a particularly contrived example. It has a highly specific user and benefit, making use of a feature already listed in another user story, but not to better effect.

Aulud reviewed (commented):
Overall good job! Like how some of the features are already work-in-progress.

lumwb commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
+|active youth in need of sustenance within a limited budget
+|get recommendations for food prices
+|be healthy physically and financially
+
```
Might be too many actors in User Stories. Consider narrowing scope of application.

lumwb commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
+|`* * *`
+|NUS student who loans money to a lot of friends
+|use the debt tracker to track and collect my interest
+|
```
Can include a (so that) field: "can collect my interest from loanees"

fabbbbbbyy commented on docs/DeveloperGuide.adoc:
```diff
@@ -616,12 +635,93 @@ _Use case ends_
+2. THRIFT processes the request and adds the tag to the specific transaction.
++
+_Use case ends_
+
```
good use case LGTM

lumwb commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
+|`* * *`
+|student staying in the Halls with many expenses
+|see all my expenses in a central spot
+|keep track of all the different expenses
```
Not specific to hall residents. Can consider making the actor generic or coming up with a hall specific user story

lumwb reviewed (commented):
Overall very detailed user stories. Good job

fabbbbbbyy commented on docs/DeveloperGuide.adoc:
```diff
@@ -616,12 +635,93 @@ _Use case ends_
+* 1a. The parameters specified by the user are not valid.
++
+[none]
+** 1a1. THRIFT displays an invalid parameter error to the user and the transaction is not updated.
```
Maybe you can further elaborate on the possible invalid parameters and give some examples.

fabbbbbbyy commented on docs/DeveloperGuide.adoc:
```diff
@@ -616,12 +635,93 @@ _Use case ends_
+_Use case ends_
+
+[discrete]
+=== Use case: Convert a value
```
I can see this as being a very useful use case for people who use your application.

lumwb commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
+== Non Functional Requirements
+
+. THRIFT should be able to run on any <<mainstream-os, mainstream OS>> as long as it has `Java 11` installed.
+. THRIFT should be able to hold up to 250 transactions without a noticeable drop in performance for typical usage.
```
Why 250 transactions? Is there a specific reason this limit is chosen. Additionally, maybe you could define "noticeable drop" in performance.

lumwb commented on docs/DeveloperGuide.adoc:
```diff
@@ -279,104 +298,446 @@ Refer to the guide <<DevOps#, here>>.
+. THRIFT should be able to run with or without internet connection.
+. THRIFT should work for a single user only.
+. THRIFT should not require user to install.
+. Features implemented should be testable using manual testing and automated testing.
```
Maybe could give some examples of automated testing used (unit testing, UI testing, etc.)

lumwb reviewed (commented):
Nice work!

fabbbbbbyy commented on docs/DeveloperGuide.adoc:
```diff
@@ -616,12 +635,93 @@ _Use case ends_
+
+1. User requests to untag a specific transaction.
+2. THRIFT processes the request and removes the tag from the specific transaction.
++
```
Maybe you could consider specifying what happens on the user's end after untagging. For example, a "successfully removed" message.

limerencee committed:
TransactionCard modifications and added new Balance UI (#89)

Added 'Balance' Ui pane, Modified TransactionCard into ExpenseTransactionCard and IncomeTransactionCard, Formatted the layout of ExpenseTransactionCard and IncomeTransactionCard

lightz96 committed:
Add undo feature for add_expense command (#90)

* Remove MessageTest

* Add dummy undocommand

* Rename persons to transactions

* Add PastUndoableCommands with the ability to store and retrieve command

* Add undo feature for add_expense command

lightz96 committed:
Change desktop app logo (#91)

limerencee committed:
Added 'add_income' command (#93)

* Added add_income command and tests

Other changes:
1. Improved the way large money values are shown: "1000000.00" - "1,000,000.00".
2. The raw transaction value ("1000000.00" instead of "1,000,000.00") is now stored into the JSON data file instead of the formatted one, to ease restoration of transaction objects upon application startup.

* Added comments for AddTransactionCommandParser class

Also changed the arguments AddTransactionCommandParser#parseTransactionDate() as it does not require ArgumentMultimap to return the TransactionDate.

lye-jw committed:
Refactor Edit to Update and modify EditCommand syntax (#94)

* Refactor instances of edit & classes to update

* Refactor Edit to Update and modify update syntax

* Modify tests to reflect changes in syntax

* Fix checkstyle issues

* Add 2 new use cases to DeveloperGuide

* Fix issues raised in review

lightz96 committed:
Undo feature for delete command (#99)

* Add undo feature for delete command

* Update the test cases

* Update javadoc

limerencee committed:
UI improvements to remove text truncation which results in the appearance of ellipses. (#101)

* Bugfix for big transaction values storing incorrectly into JSON

* Balance bar at the bottom now shows label for monthly budget.

1. Enabled text-wrap for the various labels in an Expense/IncomeTransactionCard.
2. Added a DateUtils library for ease of making Java's Date to work.
3. Adjusted the main UI for abit.

* Removed unused css styles

* Fixed a bug where there is no minWidth for Value labels, causing unsightly overflows

* Removed unused methods in DateUtils and formatted comments to comply with coding standards.

* Added tests for DateUtils class

limerencee committed:
Added Remark field into Transactions using the prefix 'r/'

1. Updated AddExpenseCommandParser, AddIncomeCommandParser and UpdateCommandParser to work with the new field.
2. Updated UI Transaction cards to display the Remark value.
3. Updated all relevant tests and made test for Remark.
4. Updated JsonAdaptedTransaction to ensure that the new field is saved and restored properly.
5. Updated UG and fixed the output for add_expense and add_income portions.

limerencee committed:
Changed command usage messages for add_expense, add_income and update

lightz96 committed:
Undo add_income command (#107)

* Add undo feature for add_income command

* Add assertion and logging

* Modify incorrect assertion

limerencee committed:
Merge branch 'master' of https://github.com/AY1920S1-CS2103T-W12-2/main into add-remark-to-transactions

limerencee committed:
Forgot to stage UpdateCommand in previous commit.

lye-jw committed:
Include original transaction in status message after update (#102)

* Refactor instances of edit & classes to update

* Refactor Edit to Update and modify update syntax

* Modify tests to reflect changes in syntax

* Fix checkstyle issues

* Add 2 new use cases to DeveloperGuide

* Fix issues raised in review

* Display original transaction of updated transaction in status message

* Update failing tests

limerencee committed:
Refactored delete command to require i/ prefix (#104)

* Refactored delete command to require i/ prefix

Also added proper tests for the new delete command.

* Fixed formatting and method name chage from commit abaf355

limerencee committed:
Merge branch 'master' of https://github.com/AY1920S1-CS2103T-W12-2/main into add-remark-to-transactions

Resolved conflict in UG.

limerencee committed:
Merge branch 'master' of https://github.com/AY1920S1-CS2103T-W12-2/main into add-remark-to-transactions

krusagiz committed:
Merge pull request #108 from limerencee/add-remark-to-transactions

Added Remark field into Transactions using the prefix 'r/'

lightz96 committed:
Redo add_expense command (#110)

* Modify the implementation and tests for undo add_expense and add_income

* Modify the model for redo feature

* Add dummy RedoCommand

* Add Redo feature for add_expense command and its corresponding tests

* Resolve issue raised

lightz96 committed:
Redo add_income command (#111)

* Add Undo feature for add_income command

* Remove unused import

* Display full transaction list after executing undo/redo command

* Resolves undo feature for delete command bug

* Updates javadoc and tests

* Add back the method getLastTransaction()

krusagiz committed:
Added Tag/Untag Command (#112)

* Tag command added

* Tag command added

* Added UntagCommand

- TagCommand modified

* Created TagCommand tests

* Created UntagCommand Tests

- Modified TagCommand tests

* Fixed checkstyle

* Changed Tag/Untag command section in UserGuide slightly

* Fixed coding standard violations

* Adjusted some method signature indentations

* Fixed JavaDocs whitespace

krusagiz committed:
Added non-alphanumeric exception catch to Tag/Untag commands (#114)

* Tag command added

* Tag command added

* Added UntagCommand

- TagCommand modified

* Created TagCommand tests

* Created UntagCommand Tests

- Modified TagCommand tests

* Fixed checkstyle

* Changed Tag/Untag command section in UserGuide slightly

* Fixed coding standard violations

* Adjusted some method signature indentations

* Fixed JavaDocs whitespace

* Added exception catch for non-alphanumeric tag names

* Fixed merge

lightz96 committed:
Redo delete command (#115)

* Update the badges

* Redo feature for delete command

lye-jw committed:
Improve UI indication for update operation (#113)

* Implement scroll to TransactionCard of transaction after update

* Add correct implementation of scroll to updated item UI feature

* Rectify failing tests and checkstyle issues

* Fix issues raised by reviews reviewed

lye-jw committed:
Fix and add new tests for Update

lightz96 committed:
Undo and redo feature for Update command (#117)

* Add undo feature for update command

* Resolve Value#equals bug

* Add redo feature for update command

* Resolves typo

krusagiz committed:
General Bugfix for Tag/Untag (#118)

* Added check for multiple index given for Tag/Untag

- Added new getSingleValue method in ArgumentMultimap
to get a singular value that is checked for extras
- Added new message format in Messages

* Changed constant name

* Fixed unused imports

* SLAP Tag/Untag Parser

* Removed unused exception catch

lye-jw committed:
Merge branch 'master' into fix-and-add-new-tests-for-update

krusagiz committed:
Merge pull request #121 from lye-jw/fix-and-add-new-tests-for-update

Fix tests and add new tests for Update

kensonoenqh committed:
Kenson list extension (#119)

* added ListCommandParser to read arguments after command "list"

* Further edit listcommand to return exceptions when bad input or wrong prefixes are used.

* Cleaned up ListCommand to make 1.3 changes easier.

* Tried to SLAP parse by creating getCommand. edited the conditions to make it more clean..

* Changed getCommand to private.

limerencee committed:
Fixed the bug where updating similar transaction would result in unintended behaviour (#132)

* Fixed a bug whereby when updating similar transactions, the first occurrence would update instead

* Added JUnit tests for new update method

limerencee committed:
Fixed a bug whereby when deleting similar transactions, the first occurrence would be deleted instead (#138)

krusagiz committed:
Fixed Tag/Untag index selection bug (#142)

lye-jw committed:
Create classes for Clone command and related JUnit tests (#141)

* Implement CloneCommand and CloneCommandParser classes

* Create CloneCommandTest and CloneCommandParser classes containing JUnit tests

* Fix issues raised by reviews & updated UserGuide

lye-jw committed:
Add more tests for Parser classes to increase code coverage (#148)

* add test cases for some parser classes

* Add more tests and fix import errors

* Fix minor Javadoc comment typo

krusagiz committed:
Added Tests for getSingleValue() (#151)

* Fixed Tag/Untag index selection bug

* Added test case for getSingleValue

* Fixed Checkstyle

* Added space inside lambda

lightz96 committed:
Undo and redo features for clone command (#156)

* Adds undo and redo feature for clone command

* Removes unused import

limerencee committed:
Budget command added allowing user to specify their budget for a specific month (#143)

* Budget command added which allows user to specify their budget for a specific month.

Month are uniquely identified by the following format: MM/yyyy.
Other changes:
1. Updated UGDG to include `budget` command.
2. Updated prefix for money from `c/COST` to `v/VALUE`, UGDG are updated to reflect this change.
3. Added classes to support `BudgetList` which stores `Budget` objects containing `Value` and `Calendar`.
4. Added JUnit test cases for `budget` command and the supporting classes like `BudgetList`, `BudgetCommand`, `BudgetCommandParser` and `ThriftTest` itself.

* Added saving/reading of monthly budgets into the data file.

1. Setting budget now saves into the JSON file, similar to how `Transaction` is stored.
2. Added JUnit tests for (de)/serializing of `Budget` objects.
3. Fixed JUnit tests for `JsonSerializableThrift`.
4. Added `StreamUtils` for converting `Iterator<T>` to `Stream<T>`. JUnit tests are also added for this new utility class.

* Fixed missing EOF in test json file

* THRIFT now shows the current month's transactions on startup.

1. New `TransactionIsInMonthYearPredicateTest` file to check if a `Transaction` belongs to that month. JUnit tests are added as well.
2. Current month's budget is shown in GUI.
3. Current month's balance is shown in GUI.

* Refreshes and displays the current month's transaction list after add/delete/update is executed

* Supported dynamic updating of displayed month's budget and balance

1. After parsing successfully either a Add/Budget/Delete/Update command, reflect the possible changes on the GUI.

* Bugfix to now include `Undo` and `Redo` commands to trigger refreshing the GUI.

* Fix checkstyle

* When Add/Budget/Delete/Update a transaction, the balance will only be updated if the transaction belongs to the current month.

* Allow budget to be $0 value and other bug fixes.

1. Clone will now update the month balance accordingly.

* Added JUnit tests to BudgetValue class

* Resolved checkstyle violations and accidental changes to DG.

* Fixed typo in DG

* Fixed checkstyle

lightz96 committed:
Fix the undo and redo features for update command (#158)

* Fixs bug #152

* Makes use of requireAllNonNull method instead

lightz96 committed:
Undo and redo features for tag command (#159)

* Adds undo feature for tag command and its test

* Adds redo feature for tag command and its test

* Refactor code

lightz96 committed:
Undo and redo feature for untag command (#160)

* Adds undo feature for untag command

* Adds redo feature for untag command

* Resolves checkstyle issues

limerencee committed:
Fixed a bug where the balance will be updated incorrectly if the transaction list is filtered (#161)

limerencee committed:
Refactored "find" command and removed "clear" command (#163)

* Removed "clear" command and refactored "find" command

`find` command now searches a `Transaction`'s `Remark` field for keyword match as well.

* Added JUnit test and fixed existing ones

krusagiz committed:
Tag/Untag Filtered List Bugfix (#164)

* Make Tag/Untag work on Filtered Lists

* Removed redundant GUI update

lye-jw committed:
Refactor Commands into scrolling and non-scrolling types (#150)

* Refactor Commands into scrolling & non-scrolling types

* Fix incorrect details in Javadoc comments

* Refactor BudgetCommand to subclass of NonScrollingCommand

* Fix checkstyle issues

limerencee committed:
Added support for scrolling into newly added Transaction if it exists in the current view (#165)

Other changes:
1. Refactored `LogicManager#execute` to introduce abstraction as there were multiple checks done to the `Command` variable which made the method long and breaks SLAP principle.

lightz96 committed:
Allow the user to view usage of a specific command (#167)

* Creates parser for HelpCommand

* Allows user to view usage of a specific command

limerencee committed:
Update DG with implementation of add transaction (#169)

* Created structure of PPP

* Updated DG with implementation on add expense/income feature

* Updated DG with implementation on add expense/income feature

* Updated DG to change diagrams and wording to THRIFT context.

krusagiz committed:
Added ConvertCommand with tests (#168)

* Added ConvertCommand with tests

- Allow for corrupted or empty currency.json

* Updated Default Currency Mappings

* Fixed coding standards

* Added DECIMAL_FORMATTER to format values

- Rearranged CliSyntax variables in alphabetical order

limerencee committed:
Prepare for the release of v1.2.1 (#172)

* Ready to create standalone JARs

Removed unnecessary `SampleDataUtil` which populates THRIFT with sample data.

* Fixed asciidoc warnings when building JAR

lightz96 committed:
Add undo and redo feature implementation in DG (#174)

* Add undo add_expense command sequence diagram

* Create images for the implementation of undo and redo feature

* Adds undo and redo implementation

* Adds sequence diagram for undo feature

* Adds sequence diagram for redo feature

* Adds design considerations

* Adds EOF in diagrams and removes redundant style.puml

* Update titles in the form of sentence case

* Edited grammar in the DG and add legends to the DG

* Update the size of the image

limerencee committed:
Updated DG diagram to improve readability (#176)

* Created structure of PPP

* Updated DG with implementation on add expense/income feature

* Updated DG with implementation on add expense/income feature

* Updated DG to change diagrams and wording to THRIFT context.

* Fixed diagrams in DG

* Fixed caption

* Adjusted colour for diagram in DG

* DG update diagram

* Fix DG diagram

kensonoenqh committed:
Kenson list command1.3 (#175)

* Rewrote listcommand to improve logic and actually throw correct exception for wrong month user input.

* edited one old test case and added some comments to understand code easier.

* Cleaned up some repetitive code snippets.

krusagiz committed:
UGDG Update (#178)

* Diagrams added

* Updated DG with Tag/Untag

* Resolved Grandma and added Diagrams

* More punctuation fixes

kensonoenqh committed:
Updated the DG to include diagram and the filtering transactions feature (#179)

* Updated DG to include 1 diagram and the filtering transactions feature

* Added my diagram to the images folder.

* Improved grammar, edited the diagram slightly.

* changed predicates to **predicates** also updated the listactivitydiagram a lil.

* Straightened the lines in my activity diagram.

* marked up predicates and edited activity diagram.

lightz96 committed:
Modify logic class diagram (#181)

* Modify logic class diagram

* Change Filtering transaction to sentence case

lye-jw committed:
Added Update command implementation to DevGuide for 1.3 (#180)

* Add implementation of Update command to DevGuide

* Add Design Considerations to Update command section in DevGuide

* Fix DevGuide casing differences in subsection titles

* Fix DevGuide Update subsection casing

liakify reviewed (commented):
1. Class diagram for Logic component looks a little cluttered, can consider splitting up into separate diagrams
2. Activity diagram for adding transactions should use a diamond box for branching instead of filling text into the box

Very comprehensive and complete, looks good overall!

cambrian-dk commented on docs/diagrams/add-transaction/AddTransactionSequence.puml:
```diff
@@ -0,0 +1,19 @@
+start
+:User executes add transaction command;
+
+if (Check input validity with AddTransactionCommandParser) then ([Invalid, throws ParseException])
```
the "throws ParseException" should belong in an action box(rounded rectangle).

cambrian-dk commented on docs/diagrams/tag-untag/TagActivity.puml:
```diff
@@ -0,0 +1,28 @@
+
+            endif
+        endwhile
+        -> [All tags checked];
```
The label for this conditional should be nearer to the start of the diamond for readability, although this could be a restriction of puml.

tcgj commented on docs/diagrams/UiClassDiagram.puml:
```diff
@@ -31,30 +33,35 @@ HiddenOutside ..> Ui
+MainWindow *-down-> "1" StatusBarFooter
 
-PersonListPanel -down-> PersonCard
+TransactionListPanel *-down-> "*" ExpenseCard
+TransactionListPanel *-down-> "*" IncomeCard
```
An expense vs income doesn't seem to be very different in implementation and display, perhaps they could use the same Card rather than having 2 different types of cards

tcgj commented on docs/diagrams/LogicClassDiagram.puml:
```diff
@@ -48,12 +63,24 @@ CliSyntax ..> Prefix
 Command .right.> Model
-note right of XYZCommand: XYZCommand = AddCommand, \nFindCommand, etc
+
+ScrollableCommand ..> UI
```
Perhaps the command shouldn't have a dependency on the UI? The UI might depend on logic, but perhaps it shouldn't be a two-way dependency

tcgj reviewed (commented):
Overall your DG looks pretty good! Almost everything was changed so that's nice.

LeowWB commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+// tag::budget[]
+=== Adding transactions
+==== Implementation
+The **add_expense**/**add_income** feature is implemented through polymorphism. `**Expense**` and `**Income**` are both subclasses of the abstract class `Transaction`.
```
Good effort to differentiate the different elements of the program through formatting! That said, I think you missed a bold on **`Transaction`**.

dalsontws commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+=== Tag/Untag feature
+==== Implementation
+
+The **tag**/**untag** command performs modifications on existing `**Transaction**` entries. Due to polymorphism, both **Expense**` and `**Income**` objects extends the abstract class `**Transaction**`, and are treated the same way in the context of the **tag**/**untag** command.
```
Formatting for Expense and Income objects are different, slightly confusing to read

LeowWB commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+
+image::add-transaction/Polymorphism.png[]
+
+Because of this polymorphism relation, many of the driver functions in `THRIFT` simply references `Transaction` and it will work for both `Expense`
```
I may be wrong, but I think you missed out some bold text in this paragraph?

dalsontws commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+.A Class Diagram illustrating the implementation of Tags inside a Transaction
+image::tag-untag/Polymorphism.png[]
+
+Due to the usage of  polymorphism, many of the driver functions in `THRIFT` uses `Transaction` for both `Expense`
```
The objects Transaction, Expense and Income are not bold in this section, compared to the previous section

LeowWB commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+which object to initialize:
+
+.Sequence diagram of how adding a new `Transaction` is processed with polymorphism
+image::add-transaction/PolymorphismCalling.png[]
```
I'm not sure if it's allowed to have two lifelines in the same column? But in any case, you should try to end the lifeline of the second `:AddTransactionCommandParser` after the destroy symbol.

LeowWB commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+about to add their own. An empty transaction list is initialized.
 
-These operations are exposed in the `Model` interface as `Model#commitAddressBook()`, `Model#undoAddressBook()` and `Model#redoAddressBook()` respectively.
+image::add-transaction/AddTransaction0.png[]
```
You may want to consider resizing this image for consistency.

LeowWB commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+The input is now checked and an attempt to parse each parameter occurs:
 
-Step 1. The user launches the application for the first time. The `VersionedAddressBook` will be initialized with the initial address book state, and the `currentStatePointer` pointing to that single address book state.
+* `**Description**` is parsed by `AddTransactionCommandParser#parseTransactionDescription(ArgumentMultimap)`
```
In my opinion, this part is unnecessarily detailed. You could probably just say that it's all parsed by various methods in `AddTransactionCommandParser`.

LeowWB commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+* `**TransactionDate**` is instantiated by `AddTransactionCommandParser#parseTransactionDate()`
 
-image::UndoRedoState0.png[]
+TIP: `**ArgumentMultimap**` is a class that stores all the parsed parameters taken from the user input.
```
Feels like TIP might not be right for this.

dalsontws commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+
+The following activity diagrams summarises what happens when the user executes a command to **tag** or **untag** a `**Transaction**`.
+
+.Activity Diagram of the Tagging process
```
Both diagrams were missing certain [else] conditions at alternate paths.

For example: Branch node for TagActivity diagram has [Some Tag(s) unchecked] condition on one path, while the other path has no condition, cannot be assumed as [else] condition.

LeowWB commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+The following activity diagram summarizes what happens when the user executes a command to add a new `**Transaction**`:
+
+.Activity diagram of adding a transaction into the transaction list
+image::add-transaction/AddTransactionSequence.png[]
```
You _may_ want to consider removing the hexagons? I'm not sure if they're allowed.

dalsontws commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+image::tag-untag/TagActivity.png[]
+
+.Activity Diagram of the Un-tagging process
+image::tag-untag/UntagActivity.png[]
```
Tag and Untag Activity diagrams are very similar, with a few changes, slightly redundant

LeowWB commented on docs/DeveloperGuide.adoc:
```diff
@@ -137,256 +167,944 @@ image:BetterModelClassDiagram.png[]
+==== Design considerations
+
+===== Aspect: Differentiating between `**Expense**` and `**Income**`
+* **Alternative 1: (current choice):** Introduce a `**Transaction**` parent class which both `**Expense**` and `**Income**`
```
This is a very minor thing, but I feel the use of bolded serif font might lead to confusion (user input?)

lightz96 committed:
Update help command (#187)

* Allow users to view usage of the commands

* Resolves grammar and typo issues

lightz96 committed:
Add undo and redo feature for budget command (#188)

* Add undo and redo feature for budget command

* Add test cases

Kalsyc commented on docs/DeveloperGuide.adoc:
```diff
@@ -115,19 +143,21 @@ NOTE: The lifeline for `DeleteCommandParser` should end at the destroy marker (X
-* stores a `UserPref` object that represents the user's preferences.
-* stores the Address Book data.
-* exposes an unmodifiable `ObservableList<Person>` that can be 'observed' e.g. the UI can be bound to this list so that the UI automatically updates when the data in the list change.
+* stores a `**UserPref**` object that represents the user's preferences.
```
I think could be formatted better and the png file should have bigger font.

lye-jw committed:
Complete Clone and Update commands (#190)

* Complete Clone and Update commands
- Clone: Implement UI scrolling to cloned item
- Update: Disallow repeated prefixes in command

* Include checks for existence of cloned and updated transactions during UI scrolling

lightz96 committed:
Modify UI layout (#193)

* Changes UI to use white theme

* Add cards to the UI

* Update the monthly budget, monthly date, montly balance in the UI

* Update filter text when user executes find or list command

* Provide up-to-date expense for current month

* Provide up-to-date income for current month

* Fix minor layout in listview cell

* Modify colour to ensure that it is consistent

* Add newline at EOF

* Add the word 'Budget' to the UI

* Add test for ModelManager

* Removes unused import

limerencee committed:
Updated DG to include budget and encryption implementation (#189)

* Created structure of PPP

* Updated DG with implementation on add expense/income feature

* Updated DG with implementation on add expense/income feature

* Updated DG to change diagrams and wording to THRIFT context.

* Fixed diagrams in DG

* Fixed caption

* Adjusted colour for diagram in DG

* DG update diagram

* Fix DG diagram

* Updated DG to include budget and encryption implementation

Also updated diagrams.

* Fixed EOF build error

* Update DG

* Corrected DG typos

* Updated DG diagrams

limerencee committed:
Fixed add scrolling issue (#195)

* Fixed add scrolling issue

* Fixed a diagram in DG

lye-jw committed:
Fix Clone scrolling issue (#194)

lightz96 committed:
Update Readme (#199)

* Change the colour of the coveralls badge

* Update Ui image and add acknowledgement

* Change Readme grammar

lightz96 committed:
Provides meaningful feedback to user for undo and redo commands (#196)

* Refactor undo and redo feature

* Provide meaningful feedback message to the user when they perform undo/redo command

* Feedback error message to user if they provide argument to exit/undo/redo command

* Resolves travis CI issues

* Resolves issue raised

* Update Ui picture

krusagiz committed:
Made Tag/Untag scroll to updated entry (#205)

krusagiz committed:
Added generation of default currency.json (#202)

* Added generation of default currency.json

* Made currency initialisation more elegant

limerencee committed:
Minor change (#208)

krusagiz committed:
Updated UGDG (#201)

limerencee committed:
Updated PPP with DG contributions (#200)

* Updated ppp with DG contributions

* Added AB3 ref

* Reduce ppp length

* Removed proposed feature for DG in PPP

* Fixed EOF error

* Grammar fixes

* Update Budget usage syntax to match DG

kensonoenqh committed:
Kenson list command1.3 (#206)

* Rewrote listcommand to improve logic and actually throw correct exception for wrong month user input.

* edited one old test case and added some comments to understand code easier.

* Cleaned up some repetitive code snippets.

* Changed to use parseDate instead of parseMonth, created calendar constructor for ListCommand to use predicate for filter. Added/edited 2 tests as well

* Changed to getSingleValue to prevent multiple m/ prefixes occuring.

* added few comments.

* updated listcommand to use updateFilteredTransactionListToCurrentMonth() instead.

* Updated UGDG to account for updated Listcommand

limerencee committed:
Updated UG and PPP (#210)

* UG with PPP update

* Fixed grammar

lightz96 committed:
Update UG on undo, redo, help, exit commands (#212)

* Update UG

* Resolves comments

lye-jw committed:
Implement Cloning multiple copies (#211)

* Implement Cloning multiple copies

* Improve design for ParserUtil#parseOccurrence() catch block

* Add tests for Occurence class and enhanced Clone

* Fix checkstyle issues

* Limit number of occurrences for clones

* Improve error messages and fix checkstyle issues

* Improve error message for Clone

* Javadoc comments corrections

kensonoenqh committed:
Kenson added tests for listcommandparsertest (#214)

* Added a few tests to ListCommandParserTest

* write more tests for listcommandparser

* Added few new tests

* added a test in listcommandparser

* Added more new tests.

* removed unneeded exception.

* add more tests

krusagiz committed:
Added dynamically coloured Tag colours (#218)

* Added dynamically coloured Tags

* Updated UIClassDiagram in DG

* Improved comments

lightz96 committed:
Update UG formatting issues (#216)

* Update the formatting issues

* Fix grammar issues

* Ensure titles are in sentence case

* Fix grammar issues

* Fix formatting issue

* Fix formatting issue

* Update UG

* Update UG

* Fix grammar issue

* Remove command syntax section from exit command

* Fix typo

* Add new entries to glossary

lightz96 committed:
Update undo and redo feature for clone command (#220)

* Update undo feature for clone command

* Update redo for clone command

* Change redo feature implementation

limerencee committed:
Updated PPP (#227)

* Update PPP

* Minor adjustments to PPP

* Fixed error in UG

* Fixed grammar and wording

* Update regex to match UG

lye-jw committed:
Update UserGuide and DeveloperGuide (#215)

* Update UserGuide for Update and Clone

* Update UserGuide with 1.3 changes

* Update UG screenshots

* Resolve review comments

kensonoenqh committed:
Added some screenshots to UG (#232)

* Edited UG

* Added screenshots to UG, and edited slightly.

* Edit errors in UG, fixed one test.

* removed tag

* Changed some command syntax for list in UG

limerencee committed:
Added assertions and updated screenshots in UG (#223)

* Added assertions and updated screenshots in UG

Also added JUnit tests for `Income` object.

* Added logging

* Merged master branch

lye-jw committed:
Improved practicality of multiple Cloning results and Added Assertions & Logging (#231)

* Improve Clone feature for more practical use

* Fix checkstyle issues

* Resolve issues raised in reviews

* Add test cases for newly discovered bugs

limerencee committed:
Fixed issues raised during mock PE (#259)

UG wording and value regex fix

lightz96 committed:
Update DG (#233)

* Migrate dev ops to DG and update the images based on review

* Migrate testing section to DG

* Migrate documentation section to DG

* Add SettingUp section

* Resolve issues raised in DG

* Change to THRIFT context

lightz96 committed:
Fix UI bug (#262)

* Fix Ui bug

* Adds FAQ to UG

lye-jw committed:
Updated Documentation for Update and Clone (#261)

* Update Clone in UG

* Update DG for update and clone

* Update screenshots in DG

* Resolve review issues

krusagiz committed:
Added Defensive Programming code (#265)

krusagiz committed:
Modified UserGuide (#267)

* Modified UserGuide

* Added FAQ

* Modified FAQ

* Refined FAQ

* Fixed Grammar errors

* Fixed capitalisation

* Sentence casing change

lightz96 committed:
Improves test cases (#266)

* Improves PastUndoCommand tests

* Update ModelClassDiagram

* Removes PastUndoableCommands from ModelManager parameter field

* Add test for NoArgumentsCommandParser

* Modify test cases

* Adds more test for undoable commands

* Change Person context to Transaction context

* Save updated diagram as image

* Update the THRIFT UI in readme and UG

* Update transaction list scrolling

* Updates UG

* Refactor CloneCommandTest

* Resolves Travis CI

* Change test method name

lye-jw committed:
Added more tests for better coverage (#268)

* Initial round of adding tests

* Add more tests

* Finish adding tests

* Add tests for NoArgumentsCommandParser

* Add one more test for Version

krusagiz committed:
Hotfixed Convert (#271)

kensonoenqh committed:
Kenson 1.4 (#272)

* Edited the error message in budget.java and the relevant error message too in my test. added 2 diagrams to DG. slightly edited ug.

* Added UGDG extracts to PPP, and edited multiple changes in DG. added test for listcommand as well.

* Edited PPP syntax notes a little.

* Minor fluency edit

* edited for fluency and to match DG extract to PPP as well

* More ppp edits and adding the features section.

* fixed error in sequence diagram.

* Edit cuz forgot to change the png file

* Minor edit to diagram and PPP

* Fixed activation bar of ListSequenceDiagram

* cropped one image to fit PPP better.

lye-jw committed:
Updated PPP (Jian Wen) (#273)

* Updated lye-jw.adoc

* Add proposed report generation feature to DG

* Fit PPP within 10 pages and slight modifications to documentation

* Address review comments

limerencee committed:
Updated PPP (#274)

* updated-screenshots

* Update description for Budget

lightz96 committed:
Added lightz96 portfolio (#275)

* Update PPP

* Update minor typo in DG

* Add EOF

* Update DG and PPP

* Resolve comments

* Resolve comments

* Add EOF

krusagiz committed:
Added PPP (#276)

* Added PPP and updated some documentation

* Updated DG

lightz96 committed:
Update manual testing and formatting in DG (#278)

* Update manual testing and formatting

* Add 1 more test case for undo and redo

kensonoenqh committed:
Dg appendix-F 1.4 (#279)

* change markup to bold and edit 1 line

lightz96 committed:
Add Unit test for PastUndoableCommand (#281)

* Create Unit test class

* Add Unit test for PastUndoableCommand

* Update test case

* Add the definition of MSS

limerencee committed:
Update appendix C, D and F (#282)

lightz96 committed:
Update PPP (#284)

krusagiz committed:
Added Tag Untag Integration Test (#286)

lye-jw committed:
Apply fixes for Appendix F of DG and PPP (#285)

krusagiz committed:
UGDG Quality Assurance (#283)

* UGDG Quality Assurance

* Included current PR into PPP

* Fixed krusagiz PPP

* Fixed krusagiz PPP

* Merged changes from lightz96 & lye-jw

* Minor QA adjustments to DG

lightz96 committed:
Update grammar and manual testing in DG (#288)

lye-jw committed:
Add null check and assertion to LogicManager (#287)

lightz96 committed:
Shift storage class diagram to image folder (#289)

lightz96 committed:
Update PPP grammar (#290)

lightz96 committed:
Update file link in DG (#291)

* Update file link

* Update Ai diagram image link

* Update file link

* Update file link
